src/tactic/interactive_expr.lean

```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
 | (of_format f) := sf.of_string $ format.to_string f
 | (compose x y) := sf.compose (sf.of_eformat x) (sf.of_eformat y)
 
+/-- `subtract x y` is some `z` when `∃ z, x = y ++ z` [fixme] this probably already exists. -/
+def subtract {α} [decidable_eq α]: (list α) → (list α) → option (list α)
+| a [] := some a -- [] ++ a = a
+| [] _ := none   -- (h::t) ++ _ ≠ []
+-- if t₂ ++ z = t₁ then (h₁ :: t₁) ++ z = (h₁ :: t₂)
+| (h₁ :: t₁) (h₂ :: t₂) := if h₁ = h₂ then subtract t₁ t₂ else none
+
+/-- Get the subexpression and sf object at the given address, if any.
+The address must point to a `sf.tag_expr` boundary. -/
+meta def sf.follow : expr.address → sf → option (expr × sf)
+| [] s := none
+| l (sf.tag_expr ea e  m) := do
+  a ← subtract l ea,
+  if a = [] then pure (e,m) else
+  sf.follow a m
+| l (sf.block _ a) := sf.follow l a
+| l (sf.highlight _ a) := sf.follow l a
+| l (sf.of_string _) := none
+| l (sf.compose a b) := sf.follow l a <|> sf.follow l b
+
+/-- Run `f` on each of the immediate child substrings of the given string. -/
+meta def sf.traverse {m} [applicative m] (f : sf → m sf) : sf → m sf
```

Comment:
There's a `traversable` type class, is that relevant here?

Reply:
Yeah it's not quite the same thing as the typeclass because the class's arg is a functor. I've seen traverse used for both of these though